### PR TITLE
Feb 18 rollup (rapid rotation fix, button state fix, backspace button)

### LINKED ISSUE #2856: Crash when the user rotates device many times and quickly on GeoWidgets
#### Software and hardware versions 
Collect v1.19.0, older versions were not verified
Android 7.0, 6.0, 5.1, 4.4

#### Problem description
Crach when the user rotates device many times and quickly on GeoWidgets

#### Steps to reproduce the problem
1. User starts form with GeoWidgets with maps appearance
2. User opens question with map
3. User rotates a few times, increasing the speed of rotation
4. App crashes
crash

#### Expected behavior
App does not crash

#### Other information 
Discovered during testing #2806

### LINKED ISSUE #2886: GeoTrace button state is inconsistent after tapping zoom
#### Software and hardware versions 
Tested and found present in Collect v1.16.0 and v1.19.0 (i.e. both before and after the geo refactor).
Tested on Android 6.0.1 on a Samsung Galaxy J2 Prime.

#### Problem description
The recording buttons are in an inconsistent state after the zoom button is pressed.

The code appears to intentionally change the recording state when the zoom button is pressed, though I can see no logical reason why we would ever want that.  Three reasons why this seems like it must be a bug and not desired behaviour:

1. The zoom button is just a way of shifting the view; it shouldn't change what happens when you record.

2. It even happens when you tap Zoom and then tap Cancel so that no zooming actually occurs.

3. The play/pause button seems clearly intended to be a toggle—either recording is active or it is paused.  But after you tap Zoom, it does not toggle on and off with two taps; it takes *three* taps to flip it on and off, and then it goes back to being a toggle.

#### Steps to reproduce the problem
1. Go to a GeoTrace question and tap "Start GeoTrace".

2. In 1.16.0, wait until the dialog appears, then tap "Zoom to current location".  In 1.19.0, tap the Zoom button.

3. Tap the "Add marker" button at the top, then tap "Start" in the dialog.  The "Add marker" button switches to a Pause button and the "Record GeoPoint button" appears.

4. Tap the Pause button at the top.   The Pause button switches back to an "Add marker" button and the "Record GeoPoint" button disappears.

5. Tap the Zoom button and go to the current location.  (In 1.16.0, you can even tap Zoom and then cancel without doing anything, and the buggy behaviour in the following steps will still occur.)

6. Tap the "Add marker" button.  The "Record GeoPoint" button appears, but the "Add marker" button _does not change_.

7. Tap the "Add marker" button again.  The "Add marker" button now switches to a Pause button, and the "Record GeoPoint" button _does not change_.

8. Tap the Pause button.  Now the Pause button switches back to an "Add marker" button, the "Record GeoPoint" button disappears, and the activity functions normally again.

#### Expected behavior
There should only be two states:
1. When recording is active, there should be a "Pause" button (and no "Add marker" button), and the "Record GeoPoint" button should be visible.
2. When recording is not active, there should be an "Add marker" button (and no "Pause" button), and the "Record GeoPoint" button should not be visible.

Tapping "Add marker" and "Pause" should toggle between states 1 and 2.  There should be no inconsistent states in between.

#### Other information 
The button-state-change calls (`setVisibility`, `setEnabled`) are scattered throughout the code, which makes it really hard to ensure that the button states are a consistent representation of the internal state.  My plan to fix this bug involves moving all these calls into one method, which will simplify the code and reduce the possibility of future bugs.

### PR DESCRIPTION
**Pull request by zestyping**
### Overview

This PR rolls up three changes:

1. Closes #2856: Crash when user rotates device many times quickly on Geo widgets
2. Closes #2886: GeoTrace button state is inconsistent after tapping Zoom
3. Adds a backspace button to GeoTrace and GeoShape to remove the last point.

### Guidance to reviewers

I recommend reviewing the commits one by one.

1. This fix is all in one commit, https://github.com/opendatakit/collect/pull/2890/commits/c97db5537b87de073cc73e720911a1e9ab4e9880.  The fix consists of defending against a null `map` or a dead `Fragment`, which can happen in certain cases when the activity is closed before initialization is complete.

2. This fix is in three commits: https://github.com/opendatakit/collect/pull/2890/commits/0f876fb62058e624faacb69a13d5015bed52a369, https://github.com/opendatakit/collect/pull/2890/commits/2e6221a71644c30091b9f2276428a5e62ff6f70e, and https://github.com/opendatakit/collect/pull/2890/commits/9e8708eb7241657d568f45b2de343b11a8de1d9b.  It will be easiest to review each one separately.

3. This adds a feature described in the [Geo improvement roadmap](https://docs.google.com/document/d/1r3wJrOVDeFQkk4kWMoW29LtLI1vbBChqJ65w1Jyg4YU/edit) as feature 2a.  Having a backspace button makes it much easier to correct mistakes, because it enables you to correct the last collected point without clearing everything and starting over.  There are two commits: https://github.com/opendatakit/collect/pull/2890/commits/c4e81f4e9416cd60fba7a1078bd27a65c4109f23 adds the button to GeoTrace, and https://github.com/opendatakit/collect/pull/2890/commits/e2d3b41a4539243117950b83c615e2d94dae80fd adds the same button to GeoShape.

#### What has been done to verify that this works as intended?
1. Manually reproduced #2856 on my Android 6.0.1 phone and in the emulator, then applied this fix and rotated the screen many times quickly in each of the three Geo widgets without a crash.  (Using the emulator makes it possible to flip the screen very rapidly by clicking the emulator's screen rotation buttons.)

2. Walked through the steps listed in #2886 and confirmed that the buttons no longer enter an incorrect state.  Also exercised the rest of GeoTrace to make sure the rest of the buttons still have their proper behaviour.

3. Manually tested adding points and using the backspace button to remove them, in both GeoTrace and GeoShape, confirming that the button removes the last point and it becomes disabled when there are no more points to remove.

#### Why is this the best possible solution? Were any other approaches considered?
1. The same change is made to all three Geo widgets (defend against null in `onStart`, `onStop`, `onSaveInstanceState`, and `initMap`).  It would be even nicer to factor out all of the duplicated logic into `BaseGeoMapActivity`, but that would involve more extensive changes and I didn't want to make this review more difficult.  Perhaps that refactoring could be done in a future PR.

2. This fix takes the approach described in #2886.  There are `setEnabled` and `setVisibility` calls scattered all around the various methods of the activity, which makes it very hard to tell whether state invariants are preserved.  This fix moves all of those calls into a single `updateUi` method—just by reading that method it is easy to prove to yourself that the manual recording button is displayed if and only if the pause button is displayed, the clear button is disabled if and only if there is nothing to clear, etc.  I would like to make the same change to `GeoPointMapActivity` and `GeoShapeActivity`; perhaps that would happen in a future PR.

3. This is pretty straightforward; it just adds an operation to the MapFragment interface, and adds a button to invoke it in both activities (GeoTrace and GeoShape).

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The user-visible bugs in #2856 and #2886 are fixed, and the new backspace feature becomes available.

#### Do we need any specific form for testing your changes? If so, please attach one.
Use the Geo widgets in the "All widgets" form.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by zestyping** [comment]
I force-pushed because I realized there was a much simpler solution to #2886, which you can now see in https://github.com/opendatakit/collect/pull/2890/commits/9e8708eb7241657d568f45b2de343b11a8de1d9b.  There's no need for any flags to keep track of whether the user has seen the settings dialog yet—we want to show the dialog whenever the user is starting a new trace, so we can just check if the trace is empty.

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java [hunk 1]
```diff
@@ -102,12 +102,12 @@
     /** While true, the point cannot be moved by dragging or long-pressing. */
     private boolean isPointLocked;
 
-    private Bundle instanceState;
+    private Bundle previousState;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: Can't we move this field to `BaseGeoMapActivity`, make protected and use in other classes instead?

Comment (zestyping) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: Sure!

Comment (zestyping) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: Done.

File: collect_app/src/main/java/org/odk/collect/android/activities/BaseGeoMapActivity.java [hunk 2]
```diff
@@ -23,22 +23,25 @@
...
-
+        previousState = savedInstanceState;
         if (savedInstanceState != null) {
             selectedLayer = savedInstanceState.getInt(MAP_LAYER_KEY);
         }
     }
 
     @Override
     protected void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/BaseGeoMapActivity.java:40]: Why did you move it up? Is it necessary? Usually, it should be at the bottom.

File: collect_app/src/main/res/layout/geotrace_layout.xml [hunk 3]
```diff
@@ -77,6 +77,14 @@
                 android:visibility="gone"
                 app:srcCompat="@drawable/ic_pause_36" />
 
+            <android.support.v7.widget.AppCompatImageButton
+                android:id="@+id/backspace"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:padding="15dp"
+                android:layout_margin="5dp"
+                android:src="@drawable/ic_backspace_arrow_selector" />
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/geotrace_layout.xml]: Why `android:src` not `app:srcCompat` like in other buttons. it's a vector file so we should rather use `app:srcCompat`?

File: collect_app/src/main/res/layout/geoshape_layout.xml [hunk 4]
```diff
@@ -42,6 +42,14 @@
         android:layout_height="wrap_content"
         android:paddingTop="6dip">
 
+        <android.support.v7.widget.AppCompatImageButton
+            android:id="@+id/backspace"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:padding="15dp"
+            android:layout_margin="5dp"
+            android:src="@drawable/ic_backspace_arrow_selector" />
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/geoshape_layout.xml]: Why `android:src` not `app:srcCompat` like in other buttons. it's a vector file so we should rather use `app:srcCompat`?

File: collect_app/src/main/java/org/odk/collect/android/activities/BaseGeoMapActivity.java (hunk 2 above)
Comment (cooperka) [reply collect_app/src/main/java/org/odk/collect/android/activities/BaseGeoMapActivity.java:40]: It shouldn't matter actually as long as there are no key collisions. If there's an intentional collision (you're modifying super's result) then you should call super first -- so I think it makes sense in general to have it first like this.

**Review by cooperka — CHANGES_REQUESTED** [review]
I love the `updateUi` consolidation.

It looks like several PNG variants of `ic_backspace_arrow` were added, can we just use a single vector XML instead? In Android Studio simply right click in folder structure > **New > Vector Asset > Clip Art** > click the little Android icon to open a picker and search for the backspace icon.

File: collect_app/src/main/res/layout/geotrace_layout.xml (hunk 3 above)
Comment (zestyping) [reply collect_app/src/main/res/layout/geotrace_layout.xml]: Oops!  Fixed.

File: collect_app/src/main/res/layout/geoshape_layout.xml (hunk 4 above)
Comment (zestyping) [reply collect_app/src/main/res/layout/geoshape_layout.xml]: Fixed, thanks!

**Comment by zestyping** [comment]
@cooperka Whoa, that's so cool!  I didn't know you could just grab vector icons like that.  Thanks, replaced!

**Comment by zestyping** [comment]
Okay, I think I've addressed all the review feedback.  Thanks for the reviews @grzesiek2010 and @cooperka!

**Review by cooperka — APPROVED** [review]

**Review by grzesiek2010 — APPROVED** [review]

**Comment by mmarciniak90** [comment]
@zestyping Backspace button to GeoTrace and GeoShape is great!
But I noticed that it works in a different way for a saved answer in GeoTrace and GeoShape.
I feel that it is not expected behavior. Let me know if I'm wrong.

| |GeoTrace|GeoShape|
| --- |---|---|
|Case 1|I added 3 points <br> I saved as Polygon<br>  I opened map again<br> Backspace button was **active** and I was able to remove points| I added 3 points <br> I opened map again<br>  Backspace button was **inactive**<br> If I added new point I was able to remove all points|
|Case 2| I added 3 points <br> I clicked on Trash icon and confirmed clearing  <br> Backspace button was **inactive**| I added 3 points <br> I clicked on Trash icon and confirmed clearing <br> Backspace button was **active**|

I did not notice more problems.

Verified cases:
* Crash when the user rotates device many times and quickly on GeoWidgets
* Pause button state after tapping zoom icon
* Button which removes the last point in GeoTrace and GeoShape
* Light/dark mode

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

@opendatakit-bot unlabel "needs testing"

**Comment by zestyping** [comment]
Thank you @mmarciniak90!  Glad you caught that problem.  I've just pushed a fix for it.

**Comment by zestyping** [comment]
Rebased cleanly on master with no changes or merge conflicts.

**Comment by mmarciniak90** [comment]
@zestyping new changes cause new problems.

Backspace button is always **active** on Android 5.1 and above for both widgets. Additionally, the trash icon is also always active. 

@opendatakit-bot unlabel "needs testing"

**Comment by zestyping** [comment]
This is very puzzling.  The only change to the geo activities is in https://github.com/opendatakit/collect/pull/2890/commits/3752395dc5c4889ac9254196d960011e6672fbd3, and that doesn't touch GeoTrace at all.

The last version @mmarciniak90 tested is `50b1d9f`.  To confirm that GeoTrace hasn't changed at all, `git diff 50b1d9f 3752395 **/GeoTraceActivity.java` or simply `git diff 50b1d9f 3752395 **/Geo*`.

**Comment by zestyping** [comment]
Okay, I've tested this and I'm pretty sure it is not these "new changes cause new problems".  The problem is unrelated to this PR; something has changed on `master` that causes all buttons to *appear* enabled.  Disabled buttons are still disabled, but they are no longer grey; everything looks black.

You can tell that the button is disabled because it doesn't get darker when you press on it.  When you press on an enabled button, the button itself (the grey square) gets a little bit darker when you press on it, and lighter when you release.

I suggest we land this PR to unblock #2891 and then separately look into fixing the general problem affecting disabled buttons on `master`.

**Comment by mmarciniak90** [comment]
@zestyping thanks for the explanation. 
Everything looks great on master branch v1.20.0-beta.0

Verified on Android: 4.2, 4.4, 6.0, 7.0, 8.1

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by zestyping** [comment]
@mmarciniak90 Wonderful, thank you!  The scrollbar issue that you noticed in #2891 is also fixed in #2904.  (That fix is not in the beta but I believe it is planned to go in for the release.)

**Pull request merged**